Multi-arch rules

r58Playz opened the pull request:
Just https://github.com/PolyMC/PolyMC/pull/514 but updated to newest develop.
Works with https://github.com/PolyMC/meta/pull/16 to provide multi-arch support.

edit: Fixes https://github.com/PolyMC/PolyMC/issues/382
~~edit 2: Fixes https://github.com/PolyMC/PolyMC/issues/765 but not really, just a workaround~~ fixed already

d-513 commented:
Please squash my commits. Dealing with GHA is painful

timoreo22 commented:
There also is a few other issues with this pr like
1) a random java file got deleted ? that's the one failing CI
2) there is still a vdf parser lib, probably shouldn't be in this pr
3) there is also a random file named share in there

All of that should probably be deleted and squashed out, otherwise the code looks fine

d-513 commented:
> There also is a few other issues with this pr like
> 
>     1. a random java file got deleted ? that's the one failing CI
> 
>     2. there is still a vdf parser lib, probably shouldn't be in this pr
> 
>     3. there is also a random file named share in there
> 
> 
> All of that should probably be deleted and squashed out, otherwise the code looks fine

yea definietly. The share is actually a symlink I did for CLION to not break, but it can be removed

r58Playz commented:
> There also is a few other issues with this pr like
> 
>     1. a random java file got deleted ? that's the one failing CI
> 
>     2. there is still a vdf parser lib, probably shouldn't be in this pr
> 
>     3. there is also a random file named share in there
> 
> 
> All of that should probably be deleted and squashed out, otherwise the code looks fine

All fixed and updated to newest upstream

d-513 commented on CMakeLists.txt:
```diff
@@ -86,7 +86,7 @@ set(Launcher_BUILD_PLATFORM "" CACHE STRING "A short string identifying the plat
 
 # The metadata server
-set(Launcher_META_URL "https://meta.polymc.org/v1/" CACHE STRING "URL to fetch Launcher's meta files from.")
+set(Launcher_META_URL "https://dada513.github.io/meta-polymc/" CACHE STRING "URL to fetch Launcher's meta files from.")
```
reset this to meta.polymc.org, this was just testing

dada513 reviewed (changes requested):
reviewing mostly own pr be like

Scrumplex commented:
@r58Playz how does this fix #765?

Scrumplex commented on launcher/CMakeLists.txt:
```diff
@@ -306,9 +306,7 @@ set(MINECRAFT_SOURCES
-    minecraft/OpSys.cpp
-    minecraft/OpSys.h
-    minecraft/ParseUtils.cpp
+        minecraft/ParseUtils.cpp
```
```suggestion
    minecraft/ParseUtils.cpp
```

DioEgizio commented:
> @r58Playz how does this fix #765?

![image](https://user-images.githubusercontent.com/83089242/172884199-5385c912-c0f1-4dc7-848e-e3bf3a3fa19a.jpg)

Scrumplex commented on launcher/SysInfo.cpp:
```diff
@@ -0,0 +1,58 @@
+#endif
+    }
+
+    QString currentArch() {
+        auto qtArch = QSysInfo::currentCpuArchitecture();
+#if defined(Q_OS_MACOS) && !defined(Q_PROCESSOR_ARM)
+        if(rosettaDetect())
+        {
+          return "arm64";
+        }
+        else
+        {
+          return "x86_64";
+        }
+#endif
+        return qtArch;
+    }
```
After thinking about this a bit: It doesn't really matter on which architecture PolyMC is running on, but actually on what architecture the JRE is running on. Shouldn't CPU arch be grabbed from there? I.e. someone running a x86-64 JRE on ARM64 should still get x86-64 natives.

r58Playz replied:
Done (finally)

Scrumplex commented on launcher/CMakeLists.txt:
```diff
@@ -628,6 +626,8 @@ SET(LAUNCHER_SOURCES
     UpdateController.h
     ApplicationMessage.h
     ApplicationMessage.cpp
+    SysInfo.h
+        SysInfo.cpp
```
```suggestion
    SysInfo.h
    SysInfo.cpp
```

Scrumplex commented:
I wouldn't call that a fix tbh.

Scrumplex commented on launcher/minecraft/Rule.cpp:
```diff
@@ -13,6 +13,24 @@
  * limitations under the License.
  */
 
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  PolyMC - Minecraft Launcher
+ *  Copyright (C) 2022 dada513 <dada513@protonmail.com>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, version 3.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
```
wrong way to do a license header.

d-513 replied:
Elaborate

Scrumplex replied:
look at the other license headers we made. You merge the original license header from MMC into our own.

Example:
```cpp
/* Copyright 2013-2021 MultiMC Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```
becomes
```cpp
// SPDX-License-Identifier: GPL-3.0-only
/*
 *  PolyMC - Minecraft Launcher
 *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, version 3.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * This file incorporates work covered by the following copyright and
 * permission notice:
 *
 *      Copyright 2013-2021 MultiMC Contributors
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 */
```

DioEgizio commented:
Yeah it's more a workaround

Scrumplex commented on launcher/minecraft/Rule.h:
```diff
@@ -13,13 +13,30 @@
  * limitations under the License.
  */
 
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  PolyMC - Minecraft Launcher
+ *  Copyright (C) 2022 dada513 <dada513@protonmail.com>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, version 3.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
```
ditto

d-513 commented:
But why download useless libraries tho? Downloading arm on x86-64 doesn’t make sense

Scrumplex commented on launcher/minecraft/Rule.h:
```diff
@@ -57,27 +74,43 @@ class Rule
+            archCorrect = arch == SysInfo::currentArch();
+        }
+        //qDebug() << "Os rule with OS required" << m_system << systemCorrect << "Arch required" << arch << archCorrect;
+        return systemCorrect && archCorrect;
```
This behavior should probably only be here if the trait `CPUArchitectureAware` is present for the version file.

See https://github.com/PolyMC/meta/pull/15#pullrequestreview-1001571508

Scrumplex commented:
> But why download useless libraries tho? Downloading arm on x86-64 doesn’t make sense

Because you can't run arm natives with a x86-64 JRE. The point is, that you can't mix CPU architectures for a running application. If the user uses a x86-64 JRE, then they *need* x86-64 natives.

Scrumplex commented on launcher/minecraft/Rule.cpp:
```diff
@@ -80,14 +113,38 @@ QJsonObject OsRule::toJson()
+    {
+        // if the arch isn't empty then it has to be archDependent right?
+        parent->setArchDependent(true);
+        archCorrect = arch == SysInfo::currentArch(settingsObjJavaArch);
```
I don't really like that we are carrying the instance settings all the way from launching to here. Maybe we could create a new data structure instead (I propose calling it `LaunchContext`) that carries data, such as java version

DioEgizio commented:
> edit 2: Fixes #765 but not really, just a workaround

that issue was fixed ages ago, just so you know

r58Playz commented:
#1033

timoreo22 commented:
Get scrumped

ryanccn commented:
Scrumped
